lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -0,0 +1,277 @@
+        if options[:device].present?
+          post[:device] = {}
+          post[:device][:manufacturer] = options[:device][:manufacturer]
+          post[:device][:model] = options[:device][:model]
+          post[:device][:promptPostalCode] = options[:device][:prompt_postal_code]
+          post[:device][:promptSecurityCode] = options[:device][:prompt_security_code]
+          post[:device][:promptStreetNumber] = options[:device][:prompt_street_number]
+          post[:device][:terminalId] = options[:device][:terminal_id]
+        end
+      end
+
+      def add_emv(post, options)
+        if options[:emv].present?
+          post[:emv] = {}
+          post[:emv][:tlvData] = options[:emv][:tlv_data]
+        end
+      end
+
+      def add_p2pe(post, options)
+        if options[:p2pe].present?
+          post[:p2pe] = {}
+          post[:p2pe][:format] = options[:p2pe][:format]
+          post[:p2pe][:ksn] = options[:p2pe][:ksn]
+        end
+      end
+
+      def add_card(post, credit_card, options)
+        post[:card] = {}
+        if options[:token].present?
+          post[:card][:token] = {}
+          post[:card][:token][:value] = options[:token][:value]
+        else
+          if credit_card.respond_to?(:number)
+            post[:card][:expirationDate] = "0#{credit_card.month}#{credit_card.year.to_s[2..4]}"
+            post[:card][:number] = credit_card.number
+            post[:card][:entryMode] = options[:entry_mode]
+            post[:card][:present] = options[:present]
+            post[:card][:securityCode] = {}
+            if options[:security_code].present?
+              post[:card][:securityCode][:indicator] = options[:security_code][:indicator]
+              post[:card][:securityCode][:value] = options[:security_code][:value]
+            end
+          end
+          end
+
+      end
+
+      def add_customer(post, options)
+        if address = options[:customer]
+          post[:customer] = {}
+          post[:customer][:addressLine1] = address[:address_line1] if address[:address_line1]
+          post[:customer][:firstName] = address[:first_name]
+          post[:customer][:lastName] = address[:last_name]
+          post[:customer][:postalCode] = address[:postal_code]
+        end
+      end
+
+      def commit(action, parameters, headers)
+        url_postfix = action == 'accesstoken' ? 'credentials' : 'transactions'
+        url = (test? ? "#{test_url}#{url_postfix}/#{action}" : "#{live_url}#{url_postfix}/#{action}")
+        response = parse(ssl_post(url, parameters.to_json, headers))
+        Response.new(
+          success_from(response),
+          message_from(response),
+          response,
+          authorization: authorization_from(response, action),
+          test: test?,
+          error_code: error_code_from(response)
+        )
+      end
+
+      def parse(body)
+        return {} if body == ''
+
+        JSON.parse(body)
+      end
+
+      def message_from(response)
```

Comment:
updated this method as per the standards

Reply:
@ajawadmirza The same for this method. `response['longText']` will return nil since the response looks like this,
```
@params=
  {"result"=>
    [{"error"=>
       {"longText"=>"No GTV PAN returned from host TOKEN01CE ENGINE16CE",
        "primaryCode"=>9101,
        "secondaryCode"=>6,
        "shortText"=>"SYSTEM ERROR"},
      "server"=>{"name"=>"UTGAPI09CE"}}]},
 @success=false,
 @test=true>
```

Reply:
yes that should be `response['result'].first['error']['longText']`. 
I didn't update the response logic but only converted them to standard methods.
@ali-hassan  will do the required changes, thanks.

Reply:
Gotcha, thanks @ajawadmirza